src/lx/print/json.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 Jamey Sharp
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include <adt/set.h>
+
+#include <fsm/fsm.h>
+#include <fsm/pred.h>
+
+#include "libfsm/internal.h" /* XXX */
+
+#include "lx/ast.h"
+#include "lx/print.h"
+
+/* TODO: centralise with libfsm */
+static unsigned int
+indexof(const struct fsm *fsm, const struct fsm_state *state)
+{
+	struct fsm_state *s;
+	unsigned int i;
+
+	assert(fsm != NULL);
+	assert(state != NULL);
+
+	for (s = fsm->sl, i = 0; s != NULL; s = s->next, i++) {
+		if (s == state) {
+			return i;
+		}
+	}
+
+	assert(!"unreached");
+	return 0;
+}
+
+static unsigned int
+zindexof(const struct ast *ast, const struct ast_zone *zone)
+{
+	struct ast_zone *z;
+	unsigned int i;
+
+	assert(ast != NULL);
+	assert(zone != NULL);
+
+	for (z = ast->zl, i = 0; z != NULL; z = z->next, i++) {
+		if (z == zone) {
+			return i;
+		}
+	}
+
+	assert(!"unreached");
+	return 0;
+}
+
+static int
+print_zone(FILE *f, const struct ast *ast, const struct ast_zone *z)
+{
+	struct fsm_state *s, *st;
+	int ret;
+
+	assert(f != NULL);
+	assert(z != NULL);
+	assert(z->fsm != NULL);
+	assert(fsm_all(z->fsm, fsm_isdfa));
+	assert(ast != NULL);
+
+	fprintf(f, "    {\n");
+	fprintf(f, "      \"initial_state\": %u,\n", indexof(z->fsm, z->fsm->start));
+	fprintf(f, "      \"states\": [\n");
+
+	for (s = z->fsm->sl; s != NULL; s = s->next) {
+		struct fsm_edge *e;
+		struct set_iter it;
+
+		fprintf(f, "        {\n");
+
+		e = set_first(s->edges, &it);
+
+		if (fsm_isend(z->fsm, s)) {
+			const struct ast_mapping *m = s->opaque;
+			assert(m != NULL);
+
+			if (m->token != NULL) {
+				fprintf(f, "          \"token\": \"$%s\",\n", m->token->s);
+			}
+			if (m->to != NULL) {
+				fprintf(f, "          \"next_zone\": %u,\n", zindexof(ast, m->to));
+			}
+
+			fprintf(f, "          \"accepts\": true%s\n", e != NULL ? "," : "");
```

Comment:
Terminology: libfsm probably should have its `.isend` field renamed to "accept" at some point. I'd like the naming to be consistent.

Likewise for your "initial_zone" field, where that's called either the starting zone (when talking about the graph), or the global zone (when talking about scope).

Reply:
And the same for "initial_state"

Reply:
And for "target", which libfsm currently calls "dst" or "to".